### ISSUE TITLE
The data is in pixel space or world coordinate space?

### ISSUE THREAD

**Issue by avijit9** [vvanirudh/srnn-pytorch#1]
If I transform the ETH provided world coordinate data to the image plane, and normalize the data between -1 to 1, I can get exactly the data what you have provided in the repo. Then if I train your model and test on one dataset, I am getting the exact values reported in the paper. However, in the paper, the distance is mentioned in the meter. I think those values is in terms of pixels. Can you please check that?

**Comment by vvanirudh** [comment]
Yes. I transform the pixel error back to real world distance error using the same homography matrix used before.

**Comment by avijit9** [comment]
But the results are not matching with paper. Could please you please upload the script you are using for converting pixel coordinate to world coordinate? I am using the following - 
```


def get_mean_error_meter(ret_nodes, nodes, assumedNodesPresent, trueNodesPresent, dataset):
    '''
    Computes average displacement error
    Parameters
    ==========

    ret_nodes : A tensor of shape pred_length x numNodes x 2
    Contains the predicted positions for the nodes

    nodes : A tensor of shape pred_length x numNodes x 2
    Contains the true positions for the nodes

    nodesPresent : A list of lists, of size pred_length
    Each list contains the nodeIDs of the nodes present at that time-step

    Returns
    =======

    Error : Mean euclidean distance between predicted trajectory and the true trajectory
    '''
    pred_length = ret_nodes.size()[0]
    error = torch.zeros(pred_length).cuda()
    counter = 0

    if dataset == 1:
        dimensions = [720, 576]

        H = torch.FloatTensor([[ 1.1048200e-02, 6.6958900e-04,  -3.3295300e+00],[ -1.5966000e-03,   1.1632400e-02,  -5.3951400e+00],[   1.1190700e-04,   1.3617400e-05,   5.4276600e-01]]).cuda()
    elif dataset == 0:
        dimensions = [640, 480]
        H = torch.FloatTensor([[2.8128700e-02,   2.0091900e-03, -4.6693600e+00],[   8.0625700e-04,   2.5195500e-02,  -5.0608800e+00],[    3.4555400e-04,   9.2512200e-05,   4.6255300e-01]]).cuda()
        
    elif dataset == 2 :
        dimensions = [720, 576]
        H = torch.FloatTensor([[  -2.5956517e-02,  -5.1572804e-18,   7.8388681e+00],[   -1.0953874e-03,   2.1664330e-02,  -1.0032272e+01],[   1.9540125e-20,   4.2171410e-19,   1.0000000e+00]]).cuda()
        
    elif dataset == 3: 
        dimensions = [720, 576]
        H = torch.FloatTensor([[  -2.5956517e-02,  -5.1572804e-18,   7.8388681e+00],[   -1.0953874e-03,   2.1664330e-02,  -1.0032272e+01],[   1.9540125e-20,   4.2171410e-19,   1.0000000e+00]]).cuda() 


    elif dataset == 4:
        dimensions = [720, 576]
        H = torch.FloatTensor([[  -2.3002776e-02,   5.3741914e-04,   8.6657256e+00],[ -5.2753792e-04,   1.9565153e-02,  -6.0889188e+00],[ 0.0000000e+00,  -0.0000000e+00,   1.0000000e+00]]).cuda()

    # ones_col = np.ones((1,))

    for tstep in range(pred_length):

        for nodeID in assumedNodesPresent:

            if nodeID not in trueNodesPresent[tstep]:
                continue

            pred_pos = ret_nodes[tstep, nodeID, :]
            true_pos = nodes[tstep, nodeID, :]

            pred_pos[0] = (pred_pos[0] + 1)/2 * dimensions[0]
            pred_pos[1] = (pred_pos[1] + 1)/2 * dimensions[1]


            true_pos[0] = (true_pos[0] + 1)/2 * dimensions[0]
            true_pos[1] = (true_pos[1] + 1)/2 * dimensions[1]
            

            ones_col  = torch.ones((1,1)).cuda()
            pred_pos =  torch.cat((pred_pos, ones_col), dim =0)
            true_pos =  torch.cat((true_pos, ones_col), dim =0)

            # pdb.set_trace()
            true_pos_met = torch.matmul(torch.t(true_pos), torch.t(H))
            pred_pos_met = torch.matmul(torch.t(pred_pos), torch.t(H))


            # pdb.set_trace()
            true_pos_met = torch.div(true_pos_met, true_pos_met[0,-1])[0][:2]
            pred_pos_met = torch.div(pred_pos_met, pred_pos_met[0,-1])[0][:2]


            # pdb.set_trace()
            error[tstep] += torch.norm(pred_pos_met - true_pos_met, p=2)
            counter += 1



        if counter != 0:
            error[tstep] = error[tstep] / counter

    # pdb.set_trace()

    return torch.mean(error)
```

**Comment by lidanzhang** [comment]
@avijit9  I am also struggling on the reproduce problem and found your code to transfer pixel coordinates to world coordinates. For ucy dataset,  I downloaded the video and annotation(vsp)files, it seems no homgraphy matrix provided. Where did you find the homography matrix for 3 ucy scenes?

**Comment by avijit9** [comment]
@lidanzhang  See AyeshaNirma's comment in this issue: https://github.com/vvanirudh/social-lstm-tf/issues/10

**Comment by lidanzhang** [comment]
I see. Thanks avijit9.

**Comment by avijit9** [comment]
@lidanzhang  let me know if you are able to reproduce the results.

**Comment by avijit9** [comment]
@lidanzhang  Sorry. I forgot to mention that the `counter`  should be inside the first for loop. You can refer the issue: https://github.com/vvanirudh/srnn-pytorch/issues/2.

For me the results are like this: 

data | 0 1 2 3 4
SRNN| 4.42 2.42 1.73 2.42 3.62 

Let me know the values you are getting.

**Comment by lidanzhang** [comment]
@avijit9  Thanks for your reminder. My results should be:
data | 0 1 2 3 4
SRNN| 4.54    4.60  3.01  3.88   5.79

**Comment by avijit9** [comment]
Do you think my conversion code is correct?

**Comment by lidanzhang** [comment]
@avijit9 , I think your code is right for eth dataset. But for ucy, the coordinate origin in the center of image. Also, in my understanding, its y-axis direction is upward.  So it needs different conversion.  
I am now calculating and visualizing the error in matlab. 

In addition, what do your think why there is a gap between our results.  How about your training parameters? In my side, I used all default hyper-parameters:
lr = 0.001;  decay_rate = 0.99...

**Comment by avijit9** [comment]
@lidanzhang  The data @vvanirudh  has used was preprocessed to address the issue you mentioned in case of UCY dataset. I have plotted the results and observed that GTs r perfectly fine. But the predictions are a bit erratic. 
I have also used the default parameters given in code as well as in the paper. I have even tried changing the hyper-params. Still no luck. I think the gap is because of the bug I mentioned in earlier comment. I am also waiting for author's reply.

**Comment by keanudicap** [comment]
I was also waiting for the conversion matrix

**Comment by keanudicap** [comment]
@avijit9 I realized that the results of samples changed quite large every time I ran it in normalized pixel space (even larger in world coordinate). Did you get the results by take average of many sample results?

**Comment by avijit9** [comment]
Yes for me also results are varying.

**Comment by keanudicap** [comment]
@avijit9 Did you also test the final displacement error? What value you got (in world space)? Also, I plotted the predicted trajectory with GT trajectory. The predicted trajectory seems quite abnormal.

**Comment by trungmanhhuynh** [comment]
I am in with this topic, let me know if you have anything new update

**Comment by tessavdheiden** [comment]
Hi,

In order to convert the predictions we need to denormalize and convert pixels to coordinates. With a slight modification to @avijit9 comment (Nov 3, 2017), I made a new script:

- First method for denormalization:
```
def mapper_to_img_coordinates(img_size, x_pos, y_pos):
    new_x = (x_pos + 1) / 2 * img_size[0]
    new_y = (y_pos + 1) / 2 * img_size[1]
    return new_x, new_y
```

- Second method for conversion pixels to meters:
```
def pixel_to_meter(pixel, H, dimensions):

    positions_temp = torch.ones(3)
    pixel_temp = (np.dot(np.linalg.inv(H), positions_temp) + 1) / 2
    z_pixel = pixel_temp[2]

    x_pixel, y_pixel = mapper_to_img_coordinates(dimensions, pixel[0], pixel[1], True)

    pixel_pos_frame_ped = np.stack((x_pixel, y_pixel, z_pixel))
    pixel_pos_frame_ped[0] = pixel_pos_frame_ped[0] * pixel_pos_frame_ped[2]
    pixel_pos_frame_ped[1] = pixel_pos_frame_ped[1] * pixel_pos_frame_ped[2]
    position_pos_frame_ped = np.dot(H, pixel_pos_frame_ped)
    return position_pos_frame_ped[0:2]
```

- Last method to compute the ADE (similar can be done for FDE):
```
def get_mean_error_in_meter(ret_nodes, nodes, assumedNodesPresent, trueNodesPresent, use_cuda, dataset):
    pred_length = ret_nodes.size()[0]
    error = torch.zeros(pred_length)
    if use_cuda:
        error = error.cuda()

    dimensions, H = get_dimension_homography(use_cuda, dataset)

    for tstep in range(pred_length):
        counter = 0
        for nodeID in assumedNodesPresent:

            if nodeID not in trueNodesPresent[tstep]:
                continue

            pred_pos = ret_nodes[tstep, nodeID, :]
            true_pos = nodes[tstep, nodeID, :]

            pred_pos = torch.from_numpy(pixel_to_meter(pred_pos, H, dimensions)).type(torch.FloatTensor)
            true_pos = torch.from_numpy(pixel_to_meter(true_pos, H, dimensions)).type(torch.FloatTensor)

            error[tstep] += torch.norm(pred_pos - true_pos)
            counter += 1

        if counter != 0:
            error[tstep] = error[tstep] / counter

    return torch.mean(error)
```

- Final numbers:
data_file = ADE /n metric = ../data/eth/univ paper = 0.390000 code pixel = 0.510 code meter = 5.422
data_file = FDE /n metric = ../data/eth/univ paper = 3.740000 code pixel = 0.576 code meter = 5.898
data_file = ADE /n metric = ../data/eth/hotel paper = 0.290000 code pixel = 0.341 code meter = 1.759
data_file = FDE /n metric = ../data/eth/hotel paper = 2.640000 code pixel = 0.378 code meter = 1.903
data_file = ADE /n metric = ../data/eth/univ paper = 0.390000 code pixel = 0.519 code meter = 5.518
data_file = FDE /n metric = ../data/eth/univ paper = 3.740000 code pixel = 0.574 code meter = 5.930
data_file = ADE /n metric = ../data/ucy/univ paper = 0.330000 code pixel = 0.565 code meter = 4.231
data_file = FDE /n metric = ../data/ucy/univ paper = 3.920000 code pixel = 0.715 code meter = 5.321
data_file = ADE /n metric = ../data/ucy/zara/zara01 paper = 0.200000 code pixel = 0.262 code meter = 2.218
data_file = FDE /n metric = ../data/ucy/zara/zara01 paper = 0.520000 code pixel = 0.219 code meter = 1.789

I also did qualitative results and made some videos. If you would like to have them, please reply to this comment or send me an e-mail.

**Comment by avijit9** [comment]
@tessavdheiden  hey, still the results are not matching with the paper. 

I would love to see the videos and analysis you have done. :)

**Comment by tessavdheiden** [comment]
@avijit9 Please give me your email address so I can send u a video with the prediction (line with x) and labels (line with circles) for different pedestrians. Here are some additional specs of the setup and video:
- Training 1000 epochs on datasets [1,2,3,4]
- Validation on dataset 0 (ETH univ).
- Best epoch was 84.

**Comment by avijit9** [comment]
@tessavdheiden  you can shoot me an email at avijitdasgupta9@gmail.com
